src/shogun/io/openml/OpenMLRun.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#include <shogun/evaluation/CrossValidationStorage.h>
+#include <shogun/io/openml/OpenMLRun.h>
+#include <shogun/io/openml/ShogunOpenML.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/machine/Machine.h>
+
+using namespace shogun;
+
+std::shared_ptr<OpenMLRun> OpenMLRun::run_model_on_task(
+    std::shared_ptr<CSGObject> model, std::shared_ptr<OpenMLTask> task)
+{
+	SG_SNOTIMPLEMENTED
+	return std::shared_ptr<OpenMLRun>();
+}
+
+std::shared_ptr<OpenMLRun> OpenMLRun::run_flow_on_task(
+    std::shared_ptr<OpenMLFlow> flow, std::shared_ptr<OpenMLTask> task)
+{
+	auto data = task->get_dataset();
+	std::shared_ptr<CFeatures> features = nullptr;
+	std::shared_ptr<CLabels> labels = nullptr;
+
+	auto model = ShogunOpenML::flow_to_model(std::move(flow), true);
+
+	labels = data->get_labels();
+	features = data->get_features(data->get_default_target_attribute());
+
+	auto storage = std::make_shared<CrossValidationStorage>();
+
+	if (task->get_split()->contains_splits())
+	{
+		auto machine = std::dynamic_pointer_cast<CMachine>(model);
+		if (!machine)
+		{
+			SG_SERROR("INTERNAL ERROR: failed to cast model to machine!\n")
+		}
+		auto train_idx = task->get_train_indices();
+		auto test_idx = task->get_test_indices();
+
+		auto xval_storage = std::make_shared<CrossValidationStorage>();
+		xval_storage->set_num_folds(task->get_num_fold());
+		xval_storage->set_num_runs(task->get_num_repeats());
+		machine->set_store_model_features(true);
```

Comment:
@karlnapf I put this line because it's in crossvalidation, but I am not sure I understand what it does? The training does fail if it isn't set because the kernel machine tries to access a support vector element that is out of range...

Reply:
kernel machines are models of the form `$f(x)=\sum_i \alpha_i k(x_i,x)`, so they can only be evaluated together with the training data `${x_i}_i$`. This means you need to keep track of that data. Right now, subset views on features modify the state of the feature object (which is stupid), so when the xvalidation validation set is evaluated, the code removes/adds another subset. This means that the data stored in the kernel machine is now different (as its state changed). The call is there to make the machine store a copy of the training data with non-zero `$\alpha_i` coefficients.

All this should be removed. Features should be immutable, subsets should be new instances that share memory with the original feature, then `store_model_features` is not needed anymore. I will have a go at this at some point soon, it has been on my list for a while

Reply:
so for now, you need this call